go/vt/topo/server.go

```diff
@@ -84,13 +84,14 @@ const (
 
 // Path for all object types.
 const (
-	CellsPath             = "cells"
-	CellsAliasesPath      = "cells_aliases"
-	KeyspacesPath         = "keyspaces"
-	ShardsPath            = "shards"
-	TabletsPath           = "tablets"
-	MetadataPath          = "metadata"
-	ExternalClusterVitess = "vitess"
+	CellsPath                = "cells"
+	CellsAliasesPath         = "cells_aliases"
+	KeyspacesPath            = "keyspaces"
+	ShardsPath               = "shards"
+	TabletsPath              = "tablets"
+	MetadataPath             = "metadata"
+	ExternalClusterVitess    = "vitess"
+	KeyspaceRoutingRulesPath = "KeyspaceRoutingRules"
```

Comment:
If this is actually the "file" aka key, and not a path aka prefix, you don't need it because there's already a constant called `KeyspaceRoutingRulesFile` in server.go.

Reply:
Can you tell me what the full path/key looks like? Is it something like `vitess/global/keyspaces/keyspace/commerce/KeyspaceRoutingRules/KeyspaceRoutingRules`? If yes, we can make it a bit more concise.

Reply:
`/vitess/global/KeyspaceRoutingRules/KeyspaceRoutingRules`

My rationale for this naming (and not using snake case here):
* Our other routing rules are global:`/vitess/global/RoutingRules` and `/vitess/global/ShardRoutingRules`
* If I used snake case it would have looked different like `/vitess/global/keyspace_routing_rules/KeyspaceRoutingRules`. So not uniform, though of course we are changing the way we are constructing the key path.
* The locks are currently like this, which I thought felt better with the PascalCase`/vitess/global/KeyspaceRoutingRules/locks/7668661805839059282`

Let me know if I should change it. Or I can also move where the consts are declared, so be more consistent with the casing ...

One other construction I was think of was `/vitess/global/KeyspaceRoutingRules/Rules` instead of `/vitess/global/KeyspaceRoutingRules/KeyspaceRoutingRules`. 




The actual keys involved including locking can be seen here: https://github.com/vitessio/vitess/pull/15807#issuecomment-2094864414

Reply:
There should be no path at all, just as there is not one for `RoutingRules` and `ShardRoutingRules`. Because they are in root and they have no variable part (like a keyspace name). You also have to lock a key, not a prefix. So the key should be: `/vitess/global/KeyspaceRoutingRules` and you lock the key (currently you're locking a prefix, which is wrong). So the lock and lease key in etcd would be (the end number is actually the lease) `/vitess/global/KeyspaceRoutingRules/locks/7668661805839059282`. Trying to lock a prefix rather than a key/file may be why we had to do some non-standard createDir stuff too. I removed my approval as I'm trending back toward my original position that much of this is non-standard and it's going to come back and bite us.

Reply:
We do lock prefixes, here is how a keyspace and its components are stored and how a keyspace is locked in the topo:
`mt` is the name of the kesyapce and locks are held at `/vitess/global/keyspaces/mt` where as we store data in `/vitess/global/keyspaces/mt/Keyspace`, `/vitess/global/keyspaces/mt/VSchema` etc.

```
% etcdctl watch "/vitess/global/keyspaces/mt" --prefix=true
PUT
/vitess/global/keyspaces/mt/Keyspace
BnoneR_vt
PUT
/vitess/global/keyspaces/mt/VSchema

PUT
/vitess/global/keyspaces/mt/locks/7668661876158266323
{
  "Action": "UpdateThrottlerConfig",
  "HostName": "RSs-Laptop.local",
  "UserName": "rohit",
  "Time": "2024-05-08T16:24:12+02:00",
  "Status": "Running"
}
PUT
/vitess/global/keyspaces/mt/Keyspace
BnoneJ
     .@R_vt
DELETE
/vitess/global/keyspaces/mt/locks/7668661876158266323

PUT
/vitess/global/keyspaces/mt/locks/7668661876158266354
{
  "Action": "CreateShard",
  "HostName": "RSs-Laptop.local",
  "UserName": "rohit",
  "Time": "2024-05-08T16:24:12+02:00",
  "Status": "Running"
}
PUT
/vitess/global/keyspaces/mt/shards/0/Shard
8
DELETE
/vitess/global/keyspaces/mt/locks/7668661876158266354

PUT
/vitess/global/keyspaces/mt/locks/7668661876158267306
{
  "Action": "RebuildKeyspace",
  "HostName": "RSs-Laptop.local",
  "UserName": "rohit",
  "Time": "2024-05-08T16:24:16+02:00",
  "Status": "Running"
}
DELETE
/vitess/global/keyspaces/mt/locks/7668661876158267306

PUT
/vitess/global/keyspaces/mt/shards/0/locks/7668661876158267488
{
  "Action": "PlannedReparentShard(zone1-0000000200, AvoidPrimary = \u003cnil\u003e)",
  "HostName": "RSs-Laptop.local",
  "UserName": "rohit",
  "Time": "2024-05-08T16:24:17+02:00",
  "Status": "Running"
}
PUT
/vitess/global/keyspaces/mt/shards/0/Shard



zone1�8B
       �������
DELETE
/vitess/global/keyspaces/mt/shards/0/locks/7668661876158267488

PUT
/vitess/global/keyspaces/mt/locks/7668661876158267757
{
  "Action": "RebuildKeyspace",
  "HostName": "RSs-Laptop.local",
  "UserName": "rohit",
  "Time": "2024-05-08T16:24:18+02:00",
  "Status": "Running"
}
DELETE
/vitess/global/keyspaces/mt/locks/7668661876158267757
```

Reply:
Ah, you're right. So we lock the full "path" rather than the "file" in this case. That also helps to explain why in the `Lock` path we check that what we're locking is a "dir" — meaning that in etcd there are other keys with this key prefix. And then what we're locking (the full path) always is, in practice, a "dir" because we require that the key/file already exists and thus the full `path` has a "child" / another key with that key prefix. That also helps to explain some related behavior around this PR that we both observed...

In this case there really is no path because it's a global key. Meaning it's just off the root of e.g. `/vitess/global` like `RoutingRules` is. In the above example, we are locking the path, but it's *the path for the specific keyspace* we want to lock: `mt`. In our case, the path would be the root itself which breaks the model.

Reply:
After talking to Rohit, I can address this. Right now we have a mix of how we are handling routing rules in the topo. So we have
`vitess/global/RoutingRules` -> table routing rules
`vitess/global/ShardRoutingRules` -> self-explanatory
`vitess/global/routing_rules/keyspace/Rules` -> newly added.

For `RoutingRules` and `ShardRoutingRules`, we are not acquiring any topo level locks at all. However, we decided that for keyspace routing rules, we do need and want to acquire a topo lock. In order to stay within vitess standard usage, that requires us to create a "directory" and lock that. Hence `routing_rules/keyspace`.

The future state would be:
`vitess/global/routing_rules/table/Rules`
`vitess/global/routing_rules/shard/Rules`
`vitess/global/routing_rules/keyspace/Rules`

and we would implement locking for table and shard level routing rules as well.
@mattlord does this make sense now?